text/0000-utf8-handling.md

```diff
@@ -0,0 +1,205 @@
+World client and server program.  This study modifies the Hello World
+client available in each language.
+
+#### Golang client
+
+With a Golang gRPC client modified to send an invalid string, e.g.,
+
+```
+	var buf bytes.Buffer
+	buf.WriteString(*name)
+	buf.Write([]byte{0xff, 0xff})
+    // ...
+	r, err := c.SayHello(ctx, &pb.HelloRequest{Name: buf.String()})
+ ```
+
+the client program exits:
+
+```
+2024/05/10 15:09:23 could not greet: rpc error: code = Internal desc = grpc: error while marshaling: string field contains invalid UTF-8
+```
+
+In this case, validation is performed on the client side.  If a client
+somehow bypasses UTF-8 validation, the server will respond with an
+error.
+
+```
+2024/05/10 15:28:13 could not greet: rpc error: code = Internal desc = grpc: error unmarshalling request: string field contains invalid UTF-8
+```
+
+It appears possible to configure gRPC and/or the protocol buffer
+library to bypass UTF-8 validation, but by default gRPC-Go will
+validate on both sides.
+
+#### Rust client
+
+With a Rust gRPC client modified to send an invalid string, e.g.,
+
+```
+    let bytes = vec![255, 255];
+    
+    let request = tonic::Request::new(HelloRequest {
+        name: unsafe { String::from_utf8_unchecked(bytes) },
+    });
+```
+
+the client program exits:
+
+```
+Error: Status { code: Internal, message: "failed to decode Protobuf message: HelloRequest.name: invalid string value: data is not UTF-8 encoded", metadata: MetadataMap { headers: {"content-type": "application/grpc", "date": "Fri, 10 May 2024 16:10:46 GMT", "content-length": "0"} }, source: None }
+```
+
+In this case, validation is only performed on the server side.
+
+#### OpenTelemetry Collector `pdata`
+
+The OpenTelemetry Collector uses Gogo protobuf package to unmarshal
+the structs that underly its `pdata` objects.  With the "gogofaster"
+compiler option is selected, its `Unmarshal()` logic does not check
+for valid UTF-8 strings.
+
+The potential for user harm is clear at this point.  gRPC libraries
+may intentionally bypass UTF-8 validation in some cases, and this may
+go unnoticed until a small amount of invalid UTF-8 data leads to data
+loss in a different part of the system where validation is strictly
+performed.  When this happens in a telemetry pipeline, it can lead
+loss of valid data that was inadvertently combined with invalid data,
+possibly through batching.
+
+### Responsibility to the user
+
+When a user uses a telemetry SDK to gain observability into their
+application, it can lead to quite the opposite result when invalid
+UTF-8 causes loss of telemetry.  When a user turns to a telemetry
+system to obseve their system and makes a UTF-8 mistake, the
+consequent total loss of telemetry is a definite bad outcome.  For the
+user to be well-served in this situation, OpenTelemetry SDKs and
+Collectors SHOULD correct invalid UTF-8 using the Unicode replacement
+character, `�` (U+FFFD) as a replacement for all invalid substrings.
```

Comment:
I just want to know the details of how the collector should deal with this requirement (it may be out of the scope of this proposal, but it is good to have a guideline for implementing this feature). Should the OpenTelemetry Collectors provide a recommended opt-in option (probably a processor) to correct invalid UTF-8?

Reply:
I finished this and moved it out of draft. I propose that Collectors should be more graceful -- receivers should allow the data to be parsed by disabling UTF-8 validation in their protocol buffer libraries (as current behavior). Exporters should expect UTF-8 validation was done by a hidden processor. There may be processors that want to have valid UTF-8 and there may be processors that output modified-invalid UTF-8, which leaves open the opportunity to re-validate UTF-8 between processors, if desired.

I'd like to see this processor automatically generated from the `pdata` definitions.

Reply:
I revised this proposal to simplify requirements for the collector.
The current PR specifies that Collectors SHOULD validate UTF-8 after they receive data, not after processors or before exporters. I have added an Alternatives Considered documenting more-safe variations and rationale for not pursuing them.

I still believe we should auto-generate validation routines from pdata defs.